### PR TITLE
Dynamically apply all hive CRDs from the operator.

### PR DESCRIPTION
**Pull request by dgoodwin**
Static list was often out of date and currently missing MachinePool.
Switch to reading all CRD files from the expected directory and applying
each.

### CODE REVIEW

File: pkg/operator/hive/hive.go [hunk 1]
```diff
@@ -5,7 +5,9 @@ import (
 	"context"
 	"crypto/md5"
 	"fmt"
+	"github.com/pkg/errors"
```
Comment (staebler) [inline pkg/operator/hive/hive.go]: This import should go down in the section with logrus.

File: pkg/operator/hive/hive.go [hunk 2]
```diff
@@ -202,6 +185,22 @@ func (r *ReconcileHiveConfig) deployHive(hLog log.FieldLogger, h *resource.Helpe
...
+	// as it requires a significant privilege escalation we would rather
+	// leave in the hands of OLM.
+	crdFiles, err := assets.AssetDir("config/crds")
+	if err != nil {
+		return errors.Wrap(err, "error listing CRD asset files")
+	}
+	for _, a := range crdFiles {
+		err = util.ApplyAsset(h, filepath.Join("config/crds", a), hLog)
+		if err != nil {
+			return err
```
Comment (staebler) [inline pkg/operator/hive/hive.go]: Wrap this error around a message indicating which crd could not be applied.

File: pkg/operator/hive/hive.go [hunk 3]
```diff
@@ -202,6 +185,22 @@ func (r *ReconcileHiveConfig) deployHive(hLog log.FieldLogger, h *resource.Helpe
...
+	// the latest in the operator to ensure updates roll out.
+	// TODO: Attempt removing this once Hive is running purely on 4.x,
+	// as it requires a significant privilege escalation we would rather
+	// leave in the hands of OLM.
+	crdFiles, err := assets.AssetDir("config/crds")
+	if err != nil {
+		return errors.Wrap(err, "error listing CRD asset files")
+	}
+	for _, a := range crdFiles {
+		err = util.ApplyAsset(h, filepath.Join("config/crds", a), hLog)
```
Comment (staebler) [inline pkg/operator/hive/hive.go]: Would you object to doing `if err := ...; err != nil {`?

File: pkg/operator/hive/hive.go [hunk 4]
```diff
@@ -227,7 +226,7 @@ func (r *ReconcileHiveConfig) deployHive(hLog log.FieldLogger, h *resource.Helpe
 	for _, isName := range removeImageSets {
 		clusterImageSet := &hivev1.ClusterImageSet{}
 		err := r.Get(context.Background(), types.NamespacedName{Name: isName}, clusterImageSet)
-		if err != nil && !errors.IsNotFound(err) {
+		if err != nil && !apierrors.IsNotFound(err) {
```
Comment (staebler) [inline pkg/operator/hive/hive.go]: Although not really related to your changes, the `err != nil` is unnecessary.

Comment (dgoodwin) [reply pkg/operator/hive/hive.go]: This boils down to a is err.(type) which makes me nervous but how about we just delete the whole block as it's ancient cleanup. :)

Comment (staebler) [reply pkg/operator/hive/hive.go]: Even better!

Although I would like to know more about what makes you nervous about removing the `err != nil`.

**Comment by dgoodwin** [comment]
Updated. We can't apply everything in rbac without careful thought as part of that is not applied by our operator rather OLM, and I believe there are random names involved, it's just these 'addon' roles we deliver. We can look at this in future but for now I'm happy just to get the main CRD pain point taken care of.

File: pkg/operator/hive/hive.go (hunk 4 above)
Comment (dgoodwin) [reply pkg/operator/hive/hive.go]: a nil err and err.(type) is not something i would ever see and assume was ok. It appears it's fine but it just looks like something that could cause a panic and crash. It reads better IMO with the != nil guard.

Comment (staebler) [reply pkg/operator/hive/hive.go]: OK. It is guaranteed from the language to not panic. It is perfectly valid to do nil.(type).

**Pull request merged**